### PR TITLE
Upgrade to gorm v2

### PR DESCRIPTION
**Pull request by krithika369**
**What this PR does / why we need it**:

Upgrading the dependency on Gorm from `github.com/jinzhu/gorm v1.9.11` to [Gorm v2](https://gorm.io/) @ `v1.24.0`. Some minor refactoring of the DB connection layer has been done.

Notable changes:
* Function `gorm.IsRecordNotFoundError` is removed. Checks replaced by: `errors.Is(err, gorm.ErrRecordNotFound)`
* Gorm's tag `association_foreignkey` is changed to `references`.
* Count() query works with `int64` now, as opposed to `int`.
* `it/database/` moved to -> `database/`. The functions in the new package are largely similar to [turing](https://github.com/caraml-dev/turing/tree/main/api/turing/database).
* `api/cmd/api/main.go`, `api/cmd/api/setup.go` - Refactored the running of DB migrations for main DB and test DB into the same function `migrateDB` in the `database/` package.
* `github.com/pilagod/gorm-cursor-paginator` library dependency has been upgraded to match the Gorm version. There are some minor changes in the API contract and behavior.
* Refactored some of the transaction rollback logic as `RollbackUnlessCommitted()` is removed in Gorm v2.
* `Select("table_name.*")`, to select all columns of a table, no longer works. However, the columns selected are automatically filtered according to the type of the variables being selected into, so I removed such `Select` calls.

**Which issue(s) this PR fixes**:

None

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Checklist**

- [ ] Added unit test, integration, and/or e2e tests
- [x] Tested locally
- [ ] Updated documentation
- [ ] Update Swagger spec if the PR introduce API changes
- [ ] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

File: api/models/prediction_job.go [hunk 1]
```diff
@@ -29,11 +29,11 @@ type PredictionJob struct {
 	// Keep metadata internal until it's needed to be exposed
 	Metadata Metadata `json:"-"`
 	// The field name has to be prefixed with the related struct name
-	// in order for gorm Preload to work with association_foreignkey
+	// in order for gorm Preload to work with references
 	VersionID       ID           `json:"version_id"`
 	VersionModelID  ID           `json:"model_id"`
 	ProjectID       ID           `json:"project_id"`
-	Environment     *Environment `json:"environment" gorm:"association_foreignkey:Name;association_autoupdate:false;association_autocreate:false;"`
```
Comment (krithika369) [inline api/models/prediction_job.go:36]: No other tables are using the `association_autoupdate:false;association_autocreate:false` tags. As such, it only affects the `Environment` and should be safe to remove.

File: api/storage/model_endpoint_storage_test.go [hunk 2]
```diff
@@ -142,7 +142,6 @@ func TestModelEndpointStorage_SaveTerminated(t *testing.T) {
 }
 
 func populateModelEndpointTable(db *gorm.DB) []*models.ModelEndpoint {
-	db = db.LogMode(true)
```
Comment (krithika369) [inline api/storage/model_endpoint_storage_test.go:145]: Not sure if we really need this. Removed.

**Comment by krithika369** [comment]
@deadlycoconuts, @ariefrahmansyah, please have a look at this PR when you have some time. Thanks.

File: api/models/version.go [hunk 3]
```diff
@@ -101,11 +101,11 @@ func (v *Version) Patch(patch *VersionPatch) error {
 	return nil
 }
 
-func (v *Version) BeforeCreate(scope *gorm.Scope) {
+func (v *Version) BeforeCreate(db *gorm.DB) error {
```
Comment (ariefrahmansyah) [inline api/models/version.go:104]: I just checked the doc and found that gorm v2 updates their hooks: https://gorm.io/docs/create.html#Create-Hooks

File: api/service/model_service.go [hunk 4]
```diff
@@ -43,8 +43,7 @@ func (service *modelsService) query() *gorm.DB {
 	return service.db.
 		Preload("Endpoints", func(db *gorm.DB) *gorm.DB {
 			return db.Preload("Environment")
-		}).
-		Select("models.*")
```
[the remark this answers is not in the document]
Comment (krithika369) [reply api/service/model_service.go:47]: Yes

File: api/models/resource_config.go [hunk 5]
```diff
@@ -27,11 +27,11 @@ type ResourceConfig struct {
 	CreatedUpdated
 }
 
-func (v *ResourceConfig) BeforeCreate(scope *gorm.Scope) {
+func (v *ResourceConfig) BeforeCreate(db *gorm.DB) error {
```
[the remark this answers is not in the document]
Comment (krithika369) [reply api/models/resource_config.go:30]: This is the reason: https://github.com/caraml-dev/merlin/pull/413#discussion_r1266321015

The interface has been updated in gorm v2.

File: api/storage/model_endpoint_storage.go [hunk 6]
```diff
@@ -78,10 +77,16 @@ func (m *modelEndpointStorage) ListModelEndpointsInProject(ctx context.Context,
 
 // Save save newModelEndpoint and its nested version endpoint objects
 func (m *modelEndpointStorage) Save(ctx context.Context, prevModelEndpoint, newModelEndpoint *models.ModelEndpoint) error {
-	tx := m.db.BeginTx(ctx, &sql.TxOptions{})
-	defer tx.RollbackUnlessCommitted()
-
+	tx := m.db.Begin()
```
[the remark this answers is not in the document]
Comment (krithika369) [reply api/storage/model_endpoint_storage.go:80]: Good point. I hadn't considered that. I've now added it where the context was previously used.

**Review by ariefrahmansyah — APPROVED** [review]
LGTM! Thanks for thoroughly upgrading our GORM dependency!

**Pull request merged**